### ISSUE TITLE
Handwriting CTC match scoring: spike paused on IAM-model digit/slash domain mismatch

### ISSUE THREAD

**Issue by hoj201** [hoj201/grader_bot#73]
## Goal

Grading currently decides correctness by sending each answer-box crop to Mathpix, which returns one decoded string with no confidence signal. Its most common failure mode is misreading a handwritten fraction bar / a "1" as "/" (or vice versa) -- see `_iter_fraction_reinterpretations` in `grading.py` (issue #39), which patches this after the fact by re-interpreting Mathpix's already-decoded text.

The goal here was a probabilistic alternative: run a pretrained handwriting CRNN over the crop, keep its raw per-timestep character probabilities (no greedy decode), and score one or more *candidate* strings against those probabilities via the CTC forward algorithm -- so a genuinely ambiguous glyph gets partial credit split between readings instead of a hard commit to one. Built as a standalone module behind a toggle next to the existing Mathpix path (not a replacement), so both could be compared side by side on real scans.

Branch: [`handwriting-ctc-match`](https://github.com/hoj201/grader_bot/tree/handwriting-ctc-match)

## What's done and verified

- **`graderbot/handwriting_match.py`**: vocabulary parsing, image preprocessing, and a hand-rolled CTC forward algorithm (log-space alpha recursion, since we don't have `torch.nn.functional.ctc_loss` available -- see below) with 16 passing pure-logic tests, including the forward algorithm checked against brute-force enumeration of every possible frame-to-label alignment for small synthetic cases.
- **Model**: [`Teklia/pylaia-iam`](https://huggingface.co/Teklia/pylaia-iam) (MIT licensed) -- a LaiaCRNN (4 conv layers + 3-layer bidirectional LSTM) pretrained on IAM English handwriting. Its 81-label vocabulary has `1` and `/` as distinct labels, which is exactly the ambiguity this was meant to score.
- **Dependency pivot, torch -> onnxruntime**: `pylaia` (the package that defines the real model) pins `torch<1.14`/Python `<3.11`, and even a plain modern `torch` publishes no macOS-x86_64 wheel for Python 3.13+ at all (Apple Silicon only from 2.6 onward) -- it cannot install on this project's Intel-Mac dev machine or its `python = "^3.13"` pin. Converted the real checkpoint to ONNX once, offline, inside a throwaway Docker container that *did* have the real `pylaia` + old torch installed, reverse-engineering the exact architecture (4 conv blocks with channels `1->12->24->48->48`, poolsizes `[2,2],[2,2],[0,0],[2,2]`, 3-layer BLSTM at 256 units) straight from the checkpoint's own pickled config rather than guessing. **Verified the ONNX export is numerically identical to the source PyTorch model** (max abs diff ~2e-5 across several input widths). Runtime dependency is `onnxruntime` (pinned `~1.23.2`, the last line with an Intel-macOS wheel), not `torch` -- also a lighter dependency for the eventual fly.io deploy.
- Also had to reverse-engineer PyLaia's exact training-time preprocessing (`ToImageTensor`: grayscale, resize to height 128 keeping aspect ratio, **invert** so ink is bright/background dark, scale to `[0,1]`) from its real source -- confirmed correct by feeding the model the word "hello" and getting back "helle. ." (near-perfect; the trailing punctuation is a known IAM quirk).

## Where it's paused, and why

Real-model testing on the actual thing we care about -- digits and `/` -- turned up a real problem:

```
rendered='1'  greedy='I'    P('1')=0.175    P('/')=0.00001
rendered='/'  greedy='I'    P('1')=0.006    P('/')=0.00000
rendered='2'  greedy='a'    P('1')=0.00000  P('/')=0.00000
```

`pylaia-iam` is trained on English prose, where digits are rare and `/` is essentially absent. It reads "1" as capital "I" (a reasonable visual confusion) and almost never proposes "/" at all regardless of what's actually drawn -- it doesn't split probability 50/50 between "1" and "/" the way we wanted, it just doesn't have much of a concept of "/" to split with.

Separately: raw CTC probabilities for a *correct* read come back small by construction (products over many timesteps/alignments) -- around 0.0001-0.2 in testing, not >0.5. A fixed threshold (as originally sketched) won't work; scoring needs to be relative (candidate vs. plausible alternatives, or vs. a blank/no-answer baseline) rather than an absolute cutoff. Not designed yet.

## Next steps (not started)

1. Decide whether to keep `pylaia-iam` and validate/refute the domain-mismatch concern against *real* scanned student answer-box crops (my testing above used synthetically rendered glyphs, which may not be representative) before writing it off, or look for a checkpoint with real exposure to digits/math notation.
2. Design the relative scoring rule for `CtcResponseScorer` (candidate-vs-alternatives or candidate-vs-blank) once (1) is resolved.
3. Wire up the actual toggle: `ResponseScorer` protocol mirroring `NameReader` (`graderbot/name_reader.py`), `MathpixResponseScorer` wrapping today's behavior unchanged, `CtcResponseScorer` new; thread through `grade_hw` -> `_grade_batch` -> `app.py`'s Grade tab, exactly parallel to the existing name-reader toggle.
4. Dockerfile/fly.toml changes to ship `models/pylaia_iam/weights.onnx` (21MB, already committed to the branch) and give the app a bit more headroom.

🤖 _Posted by Claude Code on behalf of @hoj201_

**Comment by hoj201** [comment]
PyLaia does not seem to be the correct choice for math.
